enterprise/j2ee.core/src/org/netbeans/api/j2ee/core/DeploymentDescriptors.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.api.j2ee.core;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import static java.util.Map.entry;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.BEANS;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.CONSTRAINT;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.EAR;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.VALIDATION;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.WEB;
+import static org.netbeans.api.j2ee.core.DeploymentDescriptors.Type.WEB_FRAGMENT;
+
+/**
+ * This class contains information about which schema version is used for a
+ * specific Java EE or Jakarta EE version.
+ *
+ * @author Benjamin Asbach
+ */
+public class DeploymentDescriptors {
```

Comment:
Actually I never used an `EnumMap`. I just had a brief look into it and it seems that there's no possibility to create an `EnumMap` which prevents modification. Do you think it's worth going that route without that ability?

Reply:
The nice thing about `EnumSet` is that if the enum has less than 64 elements, it is simply stored as long - its cheating since it knows all possible keys.
`EnumMap` is like an array. Both are very compact and have very fast `contains()` performance.

Compare the contains impl of `EnumMap`:
https://github.com/openjdk/jdk/blob/7d4b77ad9ee803d89eab5632f5c65ac843a68b3c/src/java.base/share/classes/java/util/EnumMap.java#L221-L223
with `MapN` which is used for `Map.of()`:
https://github.com/openjdk/jdk/blob/7d4b77ad9ee803d89eab5632f5c65ac843a68b3c/src/java.base/share/classes/java/util/ImmutableCollections.java#L1323-L1340

you should still wrap it into `Collections.unmodifiableMap(enummap)`. But you would have to write a little `ofEntries()` helper method anyway, which returns an unmodifiable view of an `EnumMap`.

Reply:
I guess I figured out to put everything into an `enum`. Not 100% if I'm happy with the syntax, but the test is still green.

Reply:
> Not 100% if I'm happy with the syntax

you could explore moving the merge logic into the constructors, so that this can be changed to:
`JAKARTA_EE_10(JAKARTA_EE_10_WEB, entry(EAR, "9")`

one constructor would have:
`DeploymentDescriptors(Entry<Type, String>... entries)`
the other:
`DeploymentDescriptors(DeploymentDescriptors other, Entry<Type, String>... entries)`

Reply:
Good idea. Applied.